### PR TITLE
feat(agents): enhance LiveKit CLI Agent with comprehensive Python ecosystem support and improved UX

### PR DESCRIPTION
**Pull request by dwayn**
Enhances the LiveKit CLI Agent's Python project detection and Docker generation capabilities, adding support for UV, Poetry, Hatch, and pip Python project types while improving the overall developer experience with better information and guidance.

Note: This PR also incorporates (and should supersede) the work from @rektdeckard in https://github.com/livekit/livekit-cli/pull/623 

  Key Features

  Enhanced Python Ecosystem Support

  - Multi-tool Detection: Added support for UV, Poetry, PDM, Hatch, and Pipenv projects
  - Intelligent Priority System: uv.lock → poetry.lock/Pipfile.lock → requirements.txt → pyproject.toml content analysis
  - TOML Parsing: Tool detection via [tool.poetry], [tool.pdm], [tool.hatch], [tool.uv] sections
  - UV-Specific Logic: Advanced detection using [dependency-groups] and UV command patterns

  Docker Generation Improvements

  - Dockerignore Integration: Recursive file traversal with pattern matching support and exclusion of invalid files for build
  - Idiomatic Docker patterns: for better build customization
  - Enhanced Documentation: Comprehensive system package guides for audio, ML, and computer vision for docker container configuration
  - Node.js Improvements: Updated templates with pnpm consistency and better practices

  UX Improvements for CLI

  - Smart Entry Point Detection: Intelligent prioritization (main.py > src/main.py > agent.py)
  - Educational Guidance: Contextual help messages and best practice recommendations
  - User-Friendly Selection: Enhanced project file selection with descriptive options

  Technical Notes

  Priority-Based Detection Logic

  1. Lock Files: uv.lock (UV) → poetry.lock/Pipfile.lock (pip-compatible)
  2. Requirements: requirements.txt (classic pip)
  3. TOML Analysis: Tool-specific sections in pyproject.toml
  4. Content Fallback: UV-specific patterns for edge cases

  Comprehensive Test Coverage

  - Test cases covering all new functionality
  - Edge case handling: Malformed TOML, complex project structures, priority conflicts
  - Multi-tool scenarios: Projects with multiple Python tools configured
  - Utility testing: Complete coverage of new helper functions

  Testing

  - ✅ All existing tests pass
  - ✅ New comprehensive test suite covers edge cases
  - ✅ Manual testing with real Python projects (UV, Poetry, PDM, pip)
  - ✅ Docker generation tested across all project types

### CODE REVIEW

File: pkg/agentfs/examples/node.Dockerfile [hunk 1]
```diff
@@ -6,20 +6,21 @@ WORKDIR /app
...
-# throw away build stage to reduce size of final image
+# Throw away build stage to reduce size of final image
 FROM base AS build
 
 RUN apt-get update -qq && apt-get install --no-install-recommends -y ca-certificates
 COPY --link . .
 
 RUN pnpm install --frozen-lockfile
-RUN npm run build
+RUN pnpm run build
```
Comment (bcherry) [inline pkg/agentfs/examples/node.Dockerfile:16]: you may also want to detect yarn (yarn.lock) or npm (package-lock.json) projects, and rename the base one here to `node.pnpm.Dockerfile`

File: pkg/agentfs/detection_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestProjectDetection(t *testing.T) {
+	// Test UV project detection with the actual agent-starter-python example
+	uvProjectPath := "/Volumes/Code/workspaces/cloud-agents/livekit-examples/agent-starter-python"
```
Comment (rektdeckard) [inline pkg/agentfs/detection_test.go:11]: Obviously tests like this won't work on anyone else's machine. But if we wanted to we could include some actual minimal examples in this repo and test them with `embed.FS`. I imagine Go's compiler is smart enough not to compile them into the binary if only referenced from test files? But I'm not sure.

File: pkg/agentfs/docker.go [hunk 3]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+		return fmt.Errorf("failed to write .dockerignore: %w", err)
 	}
 
+	fmt.Printf("\n✔ Successfully generated Docker files:\n")
+	fmt.Printf("  %s - Container build instructions\n", util.Accented("Dockerfile"))
+	fmt.Printf("  %s - Files excluded from build context\n", util.Accented(".dockerignore"))
+	fmt.Printf("\nNext steps:\n")
+	fmt.Printf("  ► Review the %s and uncomment any needed system packages\n", util.Accented("Dockerfile"))
+	fmt.Printf("  ► Build your agent: docker build -t my-agent .\n")
+	fmt.Printf("  ► Test locally: docker run my-agent\n")
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:120]: This typically won't work, as when WE run the container, we do secret injection. If we want to keep this, at the very least it will need to be `docker run --env-file .env.local my-agent`, but there's no guarantee they have the required LK secrets in their env file. Because it's iffy, I'm not sure we should say this at all?

File: pkg/agentfs/docker.go [hunk 4]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error walking directory %s: %w", dir, err)
+	}
+
+	// Convert to map for our existing prioritization logic compatibility
+	fileList := make(map[string]bool)
+	for _, file := range allFiles {
+		fileList[file] = true
 	}
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:178]: If we're gonna use a map, let's just use one from the start. Also, the "map of empty structs as a set" pattern is a tiny but very "go-pilled" optimization here 😛 

```go
allFiles := make(map[string]struct{})
```

File: pkg/agentfs/docker.go [hunk 5]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+		// Determine project type from ProjectType enum
 		var suffix string
-		switch projectType {
-		case "python":
+		if projectType.IsPython() {
 			suffix = ".py"
-		case "node":
+		} else if projectType == ProjectTypeNode {
 			suffix = ".js"
 		}
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:191]: we have `projectType.FileExt()` now

File: pkg/agentfs/docker.go [hunk 6]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+				}
 			}
 		}
 
+		// If we have a clear top choice, suggest it
+		if len(candidates) == 1 {
+			return candidates[0], nil
+		}
+
+		options := candidates
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:242]: is the copy necessary?

File: pkg/agentfs/docker.go [hunk 7]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+			case strings.Contains(option, "main."):
+				description = fmt.Sprintf("%s (common main entry point)", option)
+			case strings.Contains(option, "agent."):
+				description = fmt.Sprintf("%s (LiveKit agent)", option)
+			case strings.Contains(option, "app."):
+				description = fmt.Sprintf("%s (application entry point)", option)
+			case strings.Contains(option, "__main__.py"):
+				description = fmt.Sprintf("%s (Python module entry)", option)
+			default:
+				description = option
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:265]: I think this is pretty unnecessary

File: pkg/agentfs/docker.go [hunk 8]
```diff
@@ -51,94 +54,249 @@ func CreateDockerfile(dir string, settingsMap map[string]string) error {
...
+
+			// For Python module paths, suggest both file and module syntax
+			if projectType.IsPython() && strings.HasPrefix(option, "src/") && strings.HasSuffix(option, ".py") {
+				modulePath := strings.TrimPrefix(option, "src/")
+				modulePath = strings.TrimSuffix(modulePath, ".py")
+				modulePath = strings.ReplaceAll(modulePath, "/", ".")
+				if modulePath != "" {
+					description += fmt.Sprintf(" [can also use: src.%s]", modulePath)
+				}
+			}
```
Comment (rektdeckard) [inline pkg/agentfs/docker.go:276]: Not sure I understand what this is doing? What can you use e.g. `"src.agent.main"` for?

File: pkg/agentfs/utils.go [hunk 9]
```diff
@@ -15,23 +15,130 @@
...
+					}
+					if _, hasHatch := tool["hatch"]; hasHatch {
+						return ProjectTypePythonPip, nil
+					}
+					if _, hasUv := tool["uv"]; hasUv {
+						return ProjectTypePythonUV, nil
+					}
+				}
+
+				// Use our sophisticated UV detection for pyproject.toml files without explicit tool sections
```
Comment (rektdeckard) [inline pkg/agentfs/utils.go:118]: "sophosticated" 😆

File: pkg/agentfs/utils.go [hunk 10]
```diff
@@ -15,23 +15,130 @@
...
+// This function specifically identifies UV-based Python projects without misclassifying
+// setuptools, poetry, and other pyproject.toml-based projects as UV projects.
+func isUVByContent(content string) bool {
+	// Look for UV-specific patterns in pyproject.toml:
+	// - [dependency-groups]: UV's dependency group syntax (not used by setuptools/poetry)
+	// - "uv sync": UV command references in scripts or documentation
+	// - [tool.uv]: UV-specific tool configuration section
+	if strings.Contains(content, "[dependency-groups]") ||
+		strings.Contains(content, "uv sync") ||
+		strings.Contains(content, "[tool.uv]") {
```
Comment (rektdeckard) [inline pkg/agentfs/utils.go:141]: Don't we already check this above? Would it not also be nice to use the parsed toml doc rather than check the whole content, since we've already parsed it at this point I think this function needs a human touch to sit right with me lol.

File: pkg/util/fs_test.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	// Test 1: File exists
+	testFile := "test-file.txt"
+	os.WriteFile(filepath.Join(tempDir, testFile), []byte("test content"), 0644)
+
+	if !FileExists(tempDir, testFile) {
+		t.Errorf("Expected FileExists to return true for existing file")
+	}
+
+	// Test 2: File does not exist
+	nonExistentFile := "non-existent.txt"
```
Comment (rektdeckard) [inline pkg/util/fs_test.go:22]: I don't think we need to be testing a simple boolean wrapper around a stdlib function.

**Review by rektdeckard — COMMENTED** [review]
Some commends, but overall looks quite good. Thanks for the added user feedback, with some small style notes. Not sure about some of the entrypoint detection logic, would appreciate a little closer look.

**Review by real-danm — CHANGES_REQUESTED** [review]
let's hold off on this until after launch, i think this needs some additional work.
